apps/dataset/serializers/document_serializers.py

```diff
@@ -1064,6 +1079,28 @@ def batch_delete(self, instance: Dict, with_valid=True):
             delete_embedding_by_document_list(document_id_list)
             return True
 
+        def batch_cancel(self, instance: Dict, with_valid=True):
+            if with_valid:
+                self.is_valid(raise_exception=True)
+                BatchCancelInstanceSerializer(data=instance).is_valid(raise_exception=True)
+            document_id_list = instance.get("id_list")
+            ListenerManagement.update_status(QuerySet(Paragraph).annotate(
+                reversed_status=Reverse('status'),
+                task_type_status=Substr('reversed_status', TaskType(instance.get('type')).value,
+                                        1),
+            ).filter(task_type_status__in=[State.PENDING.value, State.STARTED.value]).filter(
+                document_id__in=document_id_list).values('id'),
+                                             TaskType(instance.get('type')),
+                                             State.REVOKE)
+            ListenerManagement.update_status(QuerySet(Document).annotate(
+                reversed_status=Reverse('status'),
+                task_type_status=Substr('reversed_status', TaskType(instance.get('type')).value,
+                                        1),
+            ).filter(task_type_status__in=[State.PENDING.value, State.STARTED.value]).filter(
+                id__in=document_id_list).values('id'),
+                                             TaskType(instance.get('type')),
+                                             State.REVOKE)
+
         def batch_edit_hit_handling(self, instance: Dict, with_valid=True):
             if with_valid:
                 BatchSerializer(data=instance).is_valid(model=Document, raise_exception=True)
```

Comment:
The provided code has several issues that need to be addressed:

1. **BatchCancelInstanceSerializer**: The validation in this serializer does not properly handle cases where the `id_list` contains non-UUID values or if elements of `id_list` do not match the expected format.

    ```python
    def is_valid(self, *, raise_exception=False):
        # Ensure all ids in id_list are valid UUIDs
        from pydantic import ValidationError
        try:
            instance_ids = list(map(uuid.UUID, self.validated_data['id_list']))
            self.data['id_list'] = instance_ids
        except ValueError as e:
            raise serializers.ValidationError(ErrMessage.uuid("id列表"))

        super().is_valid(raise_exception=True)
        _type = self.data.get('type')
        try:
            TaskType(_type)  # Validate task type enum
        except AttributeError:
            raise AppApiException(500, "未找到该任务类型")
    ```

    Additionally, ensure that `ErrorMessages.char`, `ErrorMessages.integer`, and other error messages used throughout the code are defined correctly and consistently within your project context.

2. **batch_cancel method**: Improper handling of tasks whose status might already be REVOKE or PENDING/STARTED before cancelling them could lead to redundant updates.

    ```python
    def batch_cancel(self, instance: Dict, with_valid=True):
        if with_valid:
            self.is_valid(raise_exception=True)

        document_id_list = instance.get("id_list")

        # Filter documents based on current pending or started statuses
        cancelled_docs = Paragraph.objects.filter(
            task_type_status__in=[TaskType.instance.get('type'), 'PENDING', 'STARTED'],
            document_id__in=document_id_list
        ).select_related()

        listener_update_batch(cancelled_docs, state=State.REVOKE)

        cancelled_docs = Document.objects.filter(
            task_type_status__in=[TaskType.instance.get('type'), 'PENDING', 'STARTED'],
            id__in=document_id_list
        ).select_related()

        listener_update_batch(cancelled_docs, state=State.REVOKE)
    ```
 
Consider adding more comprehensive checks to prevent unwanted operations like canceling an already canceled or ongoing transaction. For example:

```python
def listener_update_batch(queryset, start_state=None, end_state=None):
    for obj in queryset.iterator():
        current_state_code = getattr(obj, 'reversed_status')[-1].lower()
        
        # Determine whether to update based on allowed transitions
        allow_transition_from_current_to_new_state_code = True
        
        if (start_state and not allow_transition_from_current_to_new_state_code) or \
           (end_state and not allow_transition_from_current_to_new_state_code):
            
            continue
            
        new_code = get_task_type_specific_next_state(current_state_code, TaskType.instance.get('type'))
                
        LogListenerOperation(record={'operation': 'REVOKE'},
                            object=obj.id,
                            original_status=current_state_code.upper(),
                            original_state=start_state or "",
                            new_status=new_code.upper(),
                            new_state=end_state or "").execute()  
                            
        obj.reversed_status = obj.reversed_status[:-len(getattr(TaskType, str(start_state)).name)] + new_code  
        obj.state = new_code
        
        log_listener_detail(obj, action='Revoke')

        obj.save(update_fields=['reversed_status'])   
```
Ensure this additional function logic complies with your application's specific requirements for permission controls and logging. 

Additionally, review how exception propagation works during API calls – make sure it doesn't result in unexpected behavior when there’s a data integrity issue affecting multiple instances simultaneously. This might require careful attention to rollback transactions across different models or using proper error management strategies.